lib/core/controllers/colored_text_editing_controller.dart

```diff
@@ -190,6 +206,66 @@ class ColoredTextEditingController extends TextEditingController {
     );
   }
 
+  /// Filters the list of mistakes based on the changes
+  /// in the text when it is changed.
+  Iterable<Mistake> _filterMistakesOnChanged(String newText) sync* {
+    final isSelectionRangeEmpty = selection.end == selection.start;
+    final lengthDiscrepancy = newText.length - text.length;
+
+    for (final mistake in _mistakes) {
+      Mistake? newMistake;
+
+      newMistake = isSelectionRangeEmpty
+          ? _adjustMistakeOffsetWithCaretCursor(
+              mistake: mistake,
+              lengthDiscrepancy: lengthDiscrepancy,
+            )
+          : _adjustMistakeOffsetWithSelectionRange(
+              mistake: mistake,
+              lengthDiscrepancy: lengthDiscrepancy,
+            );
+
+      if (newMistake != null) yield newMistake;
+    }
+  }
+
+  /// Adjusts the mistake offset when the selection is a caret cursor.
+  Mistake? _adjustMistakeOffsetWithCaretCursor({
+    required Mistake mistake,
+    required int lengthDiscrepancy,
+  }) {
+    final mistakeRange = ClosedRange(mistake.offset, mistake.endOffset);
+    final baseOffset = selection.base.offset;
+    final shouldSkipOffsetAdjustment = mistakeRange.contains(baseOffset);
+
+    if (shouldSkipOffsetAdjustment) return null;
+
+    final shouldAdjustOffset = mistakeRange.isBeforeOrAt(baseOffset);
+    final newOffset = mistake.offset + lengthDiscrepancy;
+
+    return shouldAdjustOffset ? mistake.copyWith(offset: newOffset) : mistake;
+  }
+
+  /// Adjusts the mistake offset when the selection is a range.
+  Mistake? _adjustMistakeOffsetWithSelectionRange({
+    required Mistake mistake,
+    required int lengthDiscrepancy,
+  }) {
+    final selectionRange = ClosedRange(selection.start, selection.end);
+    final mistakeRange = ClosedRange(mistake.offset, mistake.endOffset);
+
+    final shouldSkipOffsetAdjustment =
+        selectionRange.containsRange(mistakeRange) ||
+            selectionRange.overlapsWith(mistakeRange);
+
+    if (shouldSkipOffsetAdjustment) return null;
```

Comment:
1. `ClosedRange::overlapsWith` is always true if `ClosedRange::containsRange` is true.
2. Descriptive naming
```suggestion
    final hasSelectedTextChanged = selectionRange.overlapsWith(mistakeRange);

    if (hasSelectedTextChanged) return null;
```